### PR TITLE
Fixes constructor of SubString from SubString

### PR DESCRIPTION
**Pull request by bkamins**
PR following https://discourse.julialang.org/t/do-not-allow-substring-to-grow/4438

### CODE REVIEW

**Comment by nalimilan** [comment]
I think this should rather be an error, just like when trying to index a string with an invalid index range. See how `getindex(s::String, r::UnitRange{Int})` works. Also, the check should happen in the inner constructor so that it cannot be bypassed, except by using an explicit `@inbounds` (this will require the check to be annotated with `@boundscheck`).

**Comment by davidanthoff** [comment]
I agree with @nalimilan, I think this should throw an error.

**Comment by bkamins** [comment]
Before fixing this I have one question. Do you want also:
```
julia> x=""
julia> SubString(x, 5, 10)
```
to throw an error except when annotated with `@inbounds`?

I want to make sure as I believe that in the initial design of `SubString` it was intentional to allow this.

**Comment by nalimilan** [comment]
I would say yes, but where did you read that this was supposed to work?

**Comment by bkamins** [comment]
1) @nalimilan Tests imply that. Everything starting from the line:
https://github.com/JuliaLang/julia/blob/master/test/strings/types.jl#L32
to the line:
https://github.com/JuliaLang/julia/blob/master/test/strings/types.jl#L45
should throw an error if we change this.

2) Additionally if we want to be consistent the following should be also fixed:
```
julia> x="∀"
"∀"

julia> x[2:3]
ERROR: UnicodeError: invalid character index
Stacktrace:
 [1] getindex(::String, ::UnitRange{Int64}) at .\strings\string.jl:239

julia> SubString(x, 2, 3)
""
```
In general I would recommend to have `SubString` work exactly like `function getindex(s::String, r::UnitRange{Int})` except that it does not perform a copy. If this is what is what we agree on I can go on and implement it.

3) The only question is about `@inbounds`. Because now `@inbounds` in `getindex` does not disable bounds checking - I feel that it would be safer to always check bounds in `SubString`, as with Unicode it is very error prone to assume that you guarantee that you give proper bounds.

4) Finally I would add documentation to `SubString`.

**Comment by davidanthoff** [comment]
I think that all makes sense. You could do a git blame and figure out who wrote these tests and ping that person, maybe there was a deeper rational for it.

**Comment by bkamins** [comment]
I have pushed an initial proposal how `SubString` could be improved.
@ScottPJones could you please review the assumptions how `SubString` should work, as it changes the initial design of the functionality.

**Comment by nalimilan** [comment]
Actually @ScottPJones didn't write these tests, he split tests into several files. These tests were added in https://github.com/JuliaLang/julia/pull/7145, a PR which was about introducing a more efficient implementation of `length` rather than about changing its behavior. So that doesn't say anything about the justification for that behavior.

@StefanKarpinski and @stevengj would probably know better, but copying the behavior of `getindex` sounds like a good idea to me. Regarding `@inbounds`, it's fine not to implement it for now, but I don't buy the argument that it shouldn't be added because Unicode is complex: if people use `@inbounds`, they explicitly state that they take care of using valid indices.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: It would make sense to print "$i" rather than just "i".

Comment (bkamins) [inline base/strings/types.jl]: I meant symbol `i` as it is a name of the first argument. Invalid second argument `j` is allowed (like in `getinex`).

Comment (bkamins) [inline base/strings/types.jl]: After rethinking I have hanget it to `i=$i`.

File: base/strings/types.jl
Comment (tkelman) [inline base/strings/types.jl]: it's not guaranteed that all AbstractString subtypes will have a `.len` field - that's even going to be removed from String soon

Comment (bkamins) [inline base/strings/types.jl]: @tkelman Changed to `sizeof` (I understand it will stay).

Comment (nalimilan) [inline base/strings/types.jl]: `sizeof` isn't correct here. For example, when indexing `"café"`:
```
julia> sizeof("café")
5

julia> endof("café")
4
```

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: Won't this always return `endof(s) + 1`? Why is this needed?

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: `<:Integer`

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: `j::Integer=endof(s)`.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: Same here.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: Maybe keep these examples, but with valid indices that still give the expected result?

**Comment by bkamins** [comment]
@nalimilan Thank you for the code review. I will wait with fixing it till #22548 is decided, as apart from obvious corrections your comments relate to the fact that I want to make `SubString` work like `String` with `getindex`.

Eg. currently:
```
julia> x="café"
"café"

julia> x[1:5]
"café"
```
goes through without error and that is why I have used `sizeof`.

And in general for me having consistent behavior of indexing for any `AbstractString`.

**Comment by bkamins** [comment]
Now `SubString` is defined following #22548.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: "start indexed" -> "start index". Same below.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: I'd rather keep this test, just changing the indices. It's not clear what this tested, but it must have triggered a bug in a corner case.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: Also test with start index out of bounds (both positive and negative).

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: `==isvalid(s,i)` is redundant since that's the condition. Also check that a `BoundsError` is raised when `!isvalid(s,i)`.

**Review by nalimilan — COMMENTED** [review]
I have a few comments. Looks like you'll have to dig into the bootstrap process to fix the CI failures. That's a relatively tedious work since it's hard to find out where the failures come from. Here it seems to be in the string interpolation code. You'll probably have to review all places using `SubString` and check for invalid indices.

File: base/strings/types.jl
Comment (bkamins) [inline base/strings/types.jl]: This is required as we allow any `i>j` (not necessarily valid) and return empty `SubString` then. This is required for consistency with `getindex` for `String`.

Comment (bkamins) [inline base/strings/types.jl]: OK - now I see - you meant to remove three lines with checks.

File: base/strings/types.jl
Comment (bkamins) [inline base/strings/types.jl]: Yes, but I have to retain earlier `i > j` check, as `SubString(SubString("fsfsdf", 1, 2), -10, -20)` should go through and return empty substring.

**Comment by bkamins** [comment]
@nalimilan included your comments and working through bootstrap.

File: test/strings/types.jl
Comment (tkelman) [inline test/strings/types.jl]: of a single

File: test/strings/types.jl
Comment (tkelman) [inline test/strings/types.jl]: gives equivalent results

**Comment by bkamins** [comment]
Fixed typos. Seems that some errors in Documenter.jl have to be fixed before this one is merged.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: You could continue to return `new(s, i-1, 0)` unless you've found a reason not to. I don't know what was the original justification for this behavior.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: What I meant is that you call the inner constructor below, which already checks the indices, so this is redundant (just try without it).

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: Could also include `idx = 4` here rather than in a separate line.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: Would be nice to have a comment explaining what is tested here. Same below, where some lines are commented and not others (a global comment for multiple lines is fine).

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: It's weird to reuse `str` here after tests on a different string. Maybe put these tests above those on `"∀"`?

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: Would be clearer to get rid of `str2` and use `""` everywhere.

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: "results"

**Comment by nalimilan** [comment]
Thanks. I've made a few more comments, and it seems you have not addressed a pair of old comments.

File: base/strings/types.jl
Comment (bkamins) [inline base/strings/types.jl]: The following test (which should pass) fails if I remove this line:
`@test SubString(SubString("123", 1, 2), -10, -20) == ""`
But there was an error as the line missed `return` before `SubString` so thank you for a detailed review 😄.

**Comment by bkamins** [comment]
@nalimilan Thank you. I could not find comments I have missed.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: Should be `SubString(s.string, s.offset + i - 1, s.offset)` for consistency with the other constructor (even though, as I said, I'm not sure that's the point).

**Comment by nalimilan** [comment]
I referred to comments which are still visible above:
https://github.com/JuliaLang/julia/pull/22511/files#r124203827
https://github.com/JuliaLang/julia/pull/22511/files#r124204357

**Comment by bkamins** [comment]
Unless you have something other in mind (then please let me know) I believe I have added them earlier already.

First one starts in line 188:
```
@test_throws BoundsError SubString(u, 20:30)
@test SubString(u, 20:15) == ""
@test_throws BoundsError SubString(u, -1:10)
@test SubString(u, -1, -10) == ""
@test SubString(SubString("123", 1, 2), -10, -20) == ""
```

The second test is currently reversed in lines 148-149 (I had to change some of the indices as after the PR they became invalid):
```
               SubString(s,2,11)=>"orem ipsum",
               SubString(s,15,14)=>""
```

**Comment by bkamins** [comment]
@nalimilan And I hope Documenter should have been fixed by now.
Do you know when Julia will see those changes, does it use latest Documenter on build or some tagged release?.

File: base/strings/types.jl
Comment (bkamins) [inline base/strings/types.jl]: first index: I pass ` s.offset + i` and in the inner constructor it gets converted to `s.offset + i - 1` (as you have indicated).
second index: I pass `s.offset + j` and in the inner constructor it gets converted to `0` (as you have indicated).

I have to refer to `i` and `j` so that the inner constructor can know that `i>j`.
Otherwise consider the following case `s.offset = 1`, `i = -10` and `j = -20`. Under current implementation this set of parameters will go through and generate `""` as expected (to be consistent with `getindex` for `String`). If I have changed the definition to what you propose I would pass to inner constructor `SubString(s.string, -20, 1)` which would incorrectly raise `BoundsError`.

Comment (nalimilan) [inline base/strings/types.jl]: Yes, it's converted to `0`, but that points to the beginning of the parent string, not to that of the `SubString`. Anyway I'm not sure that's of any use, so let's say the current version is OK.

**Comment by nalimilan** [comment]
I think we need a new tagged release of Documenter.

**Comment by mortenpi** [comment]
Apologies that it took so long to get these into Documenter, but once JuliaLang/METADATA.jl#10483 is merged, Documenter can be bumped to v0.11.2 which has the necessary fixes.

**Comment by tkelman** [comment]
We pin Documenter and its dependencies, so that needs to be upgraded with a separate PR - #23123

**Comment by bkamins** [comment]
I have created #23142. I hope this is what @tkelman meant.

**Comment by fredrikekre** [comment]
No, I think he meant that the Documenter upgrade was a separate PR, namely #23123 :)

**Comment by tkelman** [comment]
Right, I closed and reopened this to re-trigger CI (which runs on the merge commit) after I merged the Documenter upgrade. There were several failures that look related to this PR. Looks like someone restarted everything again so you'll have to wait, or take this branch and locally rebase it, to see what the failures all were.

**Comment by bkamins** [comment]
This is a partial fix as REPL still fails somewhere when multi-byte characters are used.

**Comment by nalimilan** [comment]
@bkamins Any chance you could finish that PR? Sorry if we haven't been responsive on the other one...

**Comment by bkamins** [comment]
@nalimilan I had problems with locating the root cause of the error and I had to put it on hold. Now I have time some to get back to it and will try to fix it.

File: base/strings/types.jl
Comment (StefanKarpinski) [inline base/strings/types.jl]: bad merge

**Comment by bkamins** [comment]
@nalimilan It seems to me that the tests are passing.
AppVeyor on ARCH=i686, JULIA_TEST_MAXRSS_MB=500 fails due to an unrelated problem (unless there is some dependency I do not understand).

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: "with the given indices" would be made a bit more specific.

File: test/strings/basic.jl
Comment (nalimilan) [inline test/strings/basic.jl]: Maybe just change this to `SubString("", 1, 1)`?

File: test/strings/search.jl
Comment (nalimilan) [inline test/strings/search.jl]: Should return `1` just like for `String`, right?

File: test/strings/types.jl
Comment (nalimilan) [inline test/strings/types.jl]: I don't understand that comment.

File: test/strings/types.jl
Comment (nalimilan) [reply test/strings/types.jl]: This comment still seems to apply.

File: test/strings/types.jl
Comment (nalimilan) [reply test/strings/types.jl]: Done below.

File: test/strings/types.jl
Comment (nalimilan) [reply test/strings/types.jl]: Done.

**Review by nalimilan — COMMENTED** [review]
Thanks, looks almost ready apart from a few details.

File: base/strings/types.jl
Comment (bkamins) [reply base/strings/types.jl]: I have made it more specific (earlier text was following documentation of `view`)

File: test/strings/basic.jl
Comment (bkamins) [reply test/strings/basic.jl]: `SubString("", 1, 1)` and `SubString("", 1, 6)` throw `BoundsError` as they are given invalid indices (like `getindex`).

File: test/strings/search.jl
Comment (bkamins) [reply test/strings/search.jl]: Should throw `BoundsError` as `SubString("",1,1)` throws such error. I have added another test case which should return `1`.

File: test/strings/types.jl
Comment (bkamins) [reply test/strings/types.jl]: I have tried to make it clearer now.

File: test/strings/types.jl
Comment (bkamins) [reply test/strings/types.jl]: The test is retained but moved to line 95.

File: test/strings/search.jl
Comment (nalimilan) [reply test/strings/search.jl]: Ah, yes, of course! But then it doesn't really make sense to pretend calling `rsearchindex` when the exception is thrown by the `SubString` constructor. Let's just remove that line and only keep the one that succeeds.

File: test/strings/types.jl
Comment (nalimilan) [reply test/strings/types.jl]: Thanks. But I think "start index" should be "end index", right?

File: test/strings/search.jl
Comment (bkamins) [reply test/strings/search.jl]: Done

File: test/strings/types.jl
Comment (bkamins) [reply test/strings/types.jl]: Fixed. Only last one tested invalid start index.

**Review by nalimilan — APPROVED** [review]
Thanks! I'd just like to get a review by either @StefanKarpinski or @stevengj and it should be good to go.

**Comment by bkamins** [comment]
Merge conflicts were introduced by #22572 - I will resolve them.

**Comment by bkamins** [comment]
I would like to ask for a merge or discard this PR before I finish implementation of #23765 as it will be using `SubString`.

File: base/regex.jl
Comment (nalimilan) [inline base/regex.jl]: Indentation should be kept as this is inside an array comprehension.

File: base/strings/types.jl
Comment (nalimilan) [inline base/strings/types.jl]: Missing backquotes around `AbstractString`, or better say "string".

**Review by nalimilan — COMMENTED** [review]
Sorry, there are two details I hadn't seen (or were introduced in the rebase).

@StefanKarpinski @stevengj Good to merge after this?

File: base/regex.jl
Comment (bkamins) [reply base/regex.jl]: fixed

File: base/strings/types.jl
Comment (bkamins) [reply base/strings/types.jl]: fixed

**Comment by bkamins** [comment]
@nalimilan Thank you for the review as the rebase was not very simple.

File: base/regex.jl
Comment (nalimilan) [inline base/regex.jl]: This one needs to be aligned with `str`. Or maybe better break after  `:` or `?`.

Comment (bkamins) [reply base/regex.jl]: I hope now it is OK (I thought I have seen the layout I have used earlier somewhere in the code base).

**Comment by bkamins** [comment]
CI failures seem to be unrelated

**Comment by bkamins** [comment]
I have performed another rebase of conflicts introduced in the meantime.
Is it good to merge?

**Comment by KristofferC** [comment]
Does this need a NEWS entry?

**Comment by bkamins** [comment]
@KristofferC I have proposed a news entry.

File: NEWS.md
Comment (nalimilan) [inline NEWS.md]: Shouldn't this be listed under "breaking changes", since it can raise an error for previously working cases?

Comment (bkamins) [reply NEWS.md]: moved, although if someone used an index outside a valid range it was probably a bug

Comment (nalimilan) [reply NEWS.md]: Sure, but that can still break code since no error was thrown before. Anyway, looks good to me now.

File: base/shell.jl
Comment (stevengj) [inline base/shell.jl]: I don't understand why these changes are being made as a part of this PR.  What does this have to do with `SubString`?

If `j` is the index of `c`, then `j-1` is actually correct for `s::String` because `c=='$'` is ASCII, by the way.

Comment (bkamins) [reply base/shell.jl]: There are two reasons:
1. if `j` is the index of `c` if `c=='$'` then `j-1` does not have to be correct. Consider `s=s="∀\$"`, then `j=4` and `j-1` is an incorrect index (`s[j-1]` throws `UnicodeError` even under stable 0.6 before the changes; the changes only make it also throw an error when range indexing is used e.g. `s[i:j-1]`).
2. It is implemented in `SubString` PR as `s` is `SubString` because it is defined as `s = lstrip(str)` in line 13 and `lstrip` returns `SubString`.

Comment (bkamins) [reply base/shell.jl]: @stevengj Is this justification OK?

Comment (stevengj) [reply base/shell.jl]: Right, I was thinking of `j+1`, not `j-1`.

**Comment by nalimilan** [comment]
@stevengj Merge now?

**Review by stevengj — APPROVED** [review]

**Comment by nalimilan** [comment]
Thanks @bkamins!

**Pull request merged**
